### PR TITLE
Minimize GUI Plugins

### PR DESCRIPTION
**Pull request by Sarath18**
Signed-off-by: Sarathkrishnan Ramesh <sarathkrishnan99@gmail.com>

Reference to issue #74

### CODE REVIEW

**Comment by Sarath18** [comment]
The PR is currently in draft state. I am working on minimizing the gui plugins when in the docked state.

**Comment by Sarath18** [comment]
Added two new states `d_collapsed` (docked collapsed) state and `f_collapsed` (floating collapsed) state and transition between these and the existing states. These states are working on config load as well.

min_plugin

File: include/ignition/gui/qml/IgnCard.qml [hunk 1]
```diff
@@ -222,12 +242,20 @@ Pane {
...
     State {
       name: "docked"
     },
-
     State {
       name: "floating"
+    },
+    // Docked collapsed state
+    State {
+      name: "d_collapsed"
```
Comment (chapulina) [inline include/ignition/gui/qml/IgnCard.qml:254]: How about being more verbose for clarity? `docked_collapsed` / `floating_collapsed` would be clearer to people not familiar with the code.

File: include/ignition/gui/qml/IgnSplit.qml [hunk 2]
```diff
@@ -328,6 +328,8 @@ SplitView {
               {
                 Layout.minimumWidth = child.Layout.minimumWidth;
               }
+              // Set child height to minimu height
```
Comment (chapulina) [inline include/ignition/gui/qml/IgnSplit.qml:331]:
```suggestion
              // Set child height to minimum height
```

File: include/ignition/gui/qml/IgnCard.qml [hunk 3]
```diff
@@ -39,6 +39,11 @@ Pane {
    */
   property bool showCloseButton: true
 
+  /**
+   * True to have a minimize button
+   */
+  property bool showMinimizeButton: true
```
Comment (chapulina) [inline include/ignition/gui/qml/IgnCard.qml:45]: How about calling it `showCollapseButton`? Just so the API is consistent. Users are able to set this, as well as the state, from config files.

**Comment by Sarath18** [comment]
> I noticed that the last docked plugin can't be minimized. Did you notice that? It looks like the internal state is collapsed, even though the icon doesn't change, because the height isn't changing.

Yes, you are right. I looked at it again and found that this case only happens when the sum of minimum heights of the plugins loaded (in split) is less than the height of the app window. If we have many expanded plugins before the last collapsed plugins, the expand and collapse work just fine. But the weird part is if you close the left split and try to expand/close the last (or any) plugins, it works as intended. Here is a gif describing the above-mentioned details.

min_bug

I am not sure why is this happening, can you help me find the issue?

**Comment by chapulina** [comment]
> can you help me find the issue?

Sure, I'll take a look this week :wink:

**Comment by Sarath18** [comment]
> Sure, I'll take a look this week :wink: 

Thank you for the help, but I believe I fixed it in the last commit 95670a7. Can you please review it?

File: include/ignition/gui/qml/IgnCard.qml [hunk 4]
```diff
@@ -376,11 +485,59 @@ Pane {
         }
         visible: card.showDockButton && !card.standalone
         onClicked: {
-          const docked = card.state === "docked"
+          const docked = (card.state === "docked" || card.state === "docked_collapsed")
           card.state = docked ? "floating" : "docked"
```
Comment (chapulina) [inline include/ignition/gui/qml/IgnCard.qml:489]: How do you feel about maintaining the collapsed / expanded state when docking / undocking?

For example, if it's currently `docked_collapsed` and I undock, I'd expect to remain collapsed, so I'd become `floating_collapsed`. The same goes when re-docking.

**Pull request merged**